### PR TITLE
RFC: Demonstrate what a function package might look like -- encoding expressions

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Builds on https://github.com/apache/arrow-datafusion/pull/8039

Demonstrates what https://github.com/apache/arrow-datafusion/issues/8045 might look like

## Rationale for this change

This PR demonstrates  what a function package API might look like by removing encoding expressions encode/decode from `BuiltInScalarFunction` enum and adding it in a separate crate (`datafusion-functions`)

## What changes are included in this PR?
1. A new `FunctionImplementation` trait and integration into `ScalarUDF` to make it easier to write ScalarUDFs;
2. a new `datafusion-functions` crate that has the implementation of `encode` and `decode`.
3. Automatically register these functions as part of `SessionState::new()`, similarly to the automatically registered `ListingTable`s
4. TODO optional enabling of functions based on feature flag

Open Questions:
1. to support the `expr_fn`s  [encode](https://docs.rs/datafusion/latest/datafusion/logical_expr/expr_fn/fn.encode.html) and [decode](https://docs.rs/datafusion/latest/datafusion/logical_expr/expr_fn/fn.decode.html), I think we will need a `Expr::ScalarFunction` call or something that can take a function by name rather than fully resolved function
2. Extract registration functions from SessionContext into their own trait / consolidate the function registry code rather than passing
around a set of HahsMaps.... And make a way to actually modify them

### CODE REVIEW

File: datafusion/expr/src/expr_fn.rs [hunk 1]
```diff
@@ -737,8 +737,9 @@ scalar_expr!(
     "converts the Unicode code point to a UTF8 character"
 );
 scalar_expr!(Digest, digest, input algorithm, "compute the binary hash of `input`, using the `algorithm`");
-scalar_expr!(Encode, encode, input encoding, "encode the `input`, using the `encoding`. encoding can be base64 or hex");
-scalar_expr!(Decode, decode, input encoding, "decode the`input`, using the `encoding`. encoding can be base64 or hex");
+// TODO make a variant of Expr that can invoke a function by name
```
Comment (alamb) [inline datafusion/expr/src/expr_fn.rs:740]: Supporting these functions will take some additional work, but I think it is doable

File: datafusion/expr/src/built_in_function.rs [hunk 2]
```diff
@@ -710,30 +704,6 @@ impl BuiltinScalarFunction {
             BuiltinScalarFunction::Digest => {
                 utf8_or_binary_to_binary_type(&input_expr_types[0], "digest")
             }
-            BuiltinScalarFunction::Encode => Ok(match input_expr_types[0] {
```
Comment (alamb) [inline datafusion/expr/src/built_in_function.rs:713]: This metadata information about the functions is now moved into the `functions/encoding/meta` module

File: datafusion/expr/src/udf.rs [hunk 3]
```diff
@@ -84,9 +94,60 @@ impl ScalarUDF {
...
+        let res = (self.return_type)(args)?;
+        Ok(res.as_ref().clone())
+    }
+    /// return the implementation of this function
+    pub fn fun(&self) -> &ScalarFunctionImplementation {
+        &self.fun
+    }
+}
+
+/// Convenience trait for implementing ScalarUDF. See [`ScalarUDF::from_impl()`]
```
Comment (alamb) [inline datafusion/expr/src/udf.rs:140]: This echo's the trait that @2010YOUY01  proposed in ) https://github.com/apache/arrow-datafusion/pull/7752, but does so in a way that is backwards compatible (makes a ScalarUDF out of the trait, to retain backwards compatibly)

File: datafusion/functions/src/encoding/meta.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+use datafusion_common::arrow::datatypes::DataType;
+use datafusion_common::{plan_err, DataFusionError, Result};
+use datafusion_expr::TypeSignature::*;
+use datafusion_expr::{ColumnarValue, FunctionImplementation, Signature, Volatility};
+use std::sync::OnceLock;
+use DataType::*;
+
+pub(super) struct EncodeFunc {}
+
+static ENCODE_SIGNATURE: OnceLock<Signature> = OnceLock::new();
```
Comment (alamb) [inline datafusion/functions/src/encoding/meta.rs:28]: Here is what `encode` and `decode` look like using the `ScalarUDF` API -- I think they are much clearer when all this type information is in one place (though I still kept it separate from the implementation to show the implementation did not change at all)

File: datafusion/functions/src/encoding/mod.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+mod inner;
+#[cfg(feature = "encoding_expressions")]
+mod meta;
+
+use crate::utils::insert;
+use datafusion_expr::ScalarUDF;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// Registers the `encode` and `decode` functions with the function registry
```
Comment (alamb) [inline datafusion/functions/src/encoding/mod.rs:28]: Here is the conditional registration of these functions based on feature flag -- there are probably nicer ways to do this but I don't think it is any worse than the current solution.

File: datafusion/functions/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+use datafusion_expr::ScalarUDF;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+pub mod encoding;
+pub mod utils;
+
+/// Registers all "built in" functions from this crate with the provided registry
+pub fn register_all(registry: &mut HashMap<String, Arc<ScalarUDF>>) {
+    encoding::register(registry);
```
Comment (alamb) [inline datafusion/functions/src/lib.rs:29]: I envision extending this list with other packages over time.

**Comment by alamb** [comment]
@2010YOUY01  and @viirya  I wonder if you have any thoughts on this approach / proposal?

File: datafusion/expr/src/udf.rs [hunk 7]
```diff
@@ -84,9 +94,60 @@ impl ScalarUDF {
...
+    /// return the return type of this function given the types of the arguments
+    pub fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+        // Old API returns an Arc of the datatype for some reason
+        let res = (self.return_type)(args)?;
+        Ok(res.as_ref().clone())
+    }
+    /// return the implementation of this function
+    pub fn fun(&self) -> &ScalarFunctionImplementation {
+        &self.fun
+    }
```
Comment (2010YOUY01) [inline datafusion/expr/src/udf.rs:137]: Is that the case this set of interfaces is internal-faced for execution, we might extend it during separating function packages?
And `trait FunctionImplementation` is the user-faced API for defining functions in separate crates

**Review by 2010YOUY01 — COMMENTED** [review]
Thank you, this looks great. I have several questions/suggestions:

1. > to support the expr_fns [encode](https://docs.rs/datafusion/latest/datafusion/logical_expr/expr_fn/fn.encode.html) and [decode](https://docs.rs/datafusion/latest/datafusion/logical_expr/expr_fn/fn.decode.html), I think we will need a Expr::ScalarFunction call or something that can take a function by name rather than fully resolved function

Now constructing an `Expr` for built-in functions is stateless (does not require context), so it's tricky to be backwards compatible for `Expr` API, the best solution I can think of is to also support initializing a UDF `Expr` with only name string, and resolve them during logical plan optimization.

2. > Extract registration functions from SessionContext into their own trait / consolidate the function registry code rather than passing
around a set of HahsMaps.... And make a way to actually modify them

It's a good idea to pack 3 HashMaps for scalar/aggr/window UDFs into a new struct like `FunctionRegistry` 👍🏼

**Comment by alamb** [comment]
> Thank you, this looks great. I have several questions/suggestions:
> 
> 1. > to support the expr_fns [encode](https://docs.rs/datafusion/latest/datafusion/logical_expr/expr_fn/fn.encode.html) and [decode](https://docs.rs/datafusion/latest/datafusion/logical_expr/expr_fn/fn.decode.html), I think we will need a Expr::ScalarFunction call or something that can take a function by name rather than fully resolved function
> 
> Now constructing an `Expr` for built-in functions is stateless (does not require context), so it's tricky to be backwards compatible for `Expr` API, the best solution I can think of is to also support initializing a UDF `Expr` with only name string, and resolve them during logical plan optimization.

Yes, I agree this approach is the best I can come up with. 

> 2. > Extract registration functions from SessionContext into their own trait / consolidate the function registry code rather than passing
>    > around a set of HahsMaps.... And make a way to actually modify them
> 
> It's a good idea to pack 3 HashMaps for scalar/aggr/window UDFs into a new struct like `FunctionRegistry` 👍🏼

👍  Unfortunately that name is already taken :) Maybe `MemoryFunctionRegistry` 🤔

File: datafusion/execution/src/registry.rs [hunk 8]
```diff
@@ -34,6 +34,17 @@ pub trait FunctionRegistry {
 
     /// Returns a reference to the udwf named `name`.
     fn udwf(&self, name: &str) -> Result<Arc<WindowUDF>>;
+
+    /// Registers a new `ScalarUDF`, returning any previously registered
```
Comment (alamb) [inline datafusion/execution/src/registry.rs:38]: this is a new proposed API -- to allow registering new scalar UDFs with a FunctionRegistry.

File: datafusion/expr/src/expr_fn.rs [hunk 9]
```diff
@@ -1228,30 +1229,4 @@ mod test {
             unreachable!();
         }
     }
-
-    #[test]
-    fn encode_function_definitions() {
```
Comment (alamb) [inline datafusion/expr/src/expr_fn.rs:1233]: I don't think these tests add a lot. A better test is to actually call `encode()` via the expr API / dataframe tests. TODO review if this is done somewhere

File: datafusion/functions/src/encoding/mod.rs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+mod inner;
+
+use datafusion_expr::{Expr, ScalarUDF};
+use std::sync::{Arc, OnceLock};
+
+pub mod expr_fn {
+    use super::*;
+    /// Return encode(arg)
+    pub fn encode(args: Vec<Expr>) -> Expr {
```
Comment (alamb) [inline datafusion/functions/src/encoding/mod.rs:26]: here are the new expr_fn implementations

**Pull request closed without merging**